src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -52,36 +51,73 @@ public static ImmutableArray<BasicBlock> Create(IBlockOperation body)
             return blocks.ToImmutableAndFree();
         }
 
+        /// <summary>
+        /// Do a pass to eliminate blocks without statements that can be merged with predecessor(s).
+        /// </summary>
+        /// <param name="blocks"></param>
         private static void Pack(ArrayBuilder<BasicBlock> blocks)
         {
             int count = blocks.Count - 1;
             for (int i = 1; i < count; i++)
             {
                 BasicBlock block = blocks[i];
                 Debug.Assert(block.Next != null);
-                if (block.Statements.IsEmpty && block.Conditional.Value == null)
+                if (block.Statements.IsEmpty)
                 {
                     BasicBlock next = block.Next;
-                    foreach (BasicBlock predecessor in block.Predecessors)
+
+                    if (block.Conditional.Value == null)
                     {
-                        if (predecessor.Next == block)
+                        if (block.Next != block)
                         {
-                            predecessor.Next = next;
-                            next.AddPredecessor(predecessor);
+                            foreach (BasicBlock predecessor in block.Predecessors)
+                            {
+                                if (predecessor.Next == block)
+                                {
+                                    predecessor.Next = next;
+                                    next.AddPredecessor(predecessor);
+                                }
+
+                                (IOperation condition, ConditionalBranchKind kind, BasicBlock destination) = predecessor.Conditional;
+                                if (destination == block)
+                                {
+                                    predecessor.Conditional = (condition, kind, next);
+                                    next.AddPredecessor(predecessor);
+                                }
+                            }
+
+                            next.RemovePredecessor(block);
+
+                            blocks.RemoveAt(i);
+                            i--;
+                            count--;
```

Comment:
Consider removing `count` and looping until `blocks.Count` instead. Or use `continue;` in code paths that do not requiring updating `blocks`, so that `next.RemovePredecessor(block); blocks.RemoveAt(i); i--;` is only needed in one location.

Reply:
> Or use continue; in code paths that do not requiring updating blocks, so that next.RemovePredecessor(block); blocks.RemoveAt(i); i--; is only needed in one location.

I don't think this will make code clearer. Especially that predecessors are removed from two distinct blocks in one case.